base/server/python/pki/server/cli/plugin.py

```diff
@@ -0,0 +1,762 @@
+
+            elif o == '--help':
+                self.print_help()
+                sys.exit()
+
+            elif o in ('-t', '--type'):
+                plugin_type = a
+
+            elif o in ('-c', '--class'):
+                plugin_class = a
+            else:
+                print('ERROR: unknown option ' + o)
+                self.print_help()
+                sys.exit(1)
+
+        if not plugin_class:
+            print("ERROR: Plugin class required.")
+            sys.exit(1)
+
+        if not plugin_type:
+            print("ERROR: Plugin type required")
+            sys.exit(1)
+
+        instance = pki.server.PKIInstance(instance_name)
+        if not instance.is_valid():
+            print('ERROR: Invalid instance %s.' % instance_name)
+            sys.exit(1)
+
+        instance.load()
+
+        subsystem_name = self.parent.parent.parent.name
+        subsystem = instance.get_subsystem(subsystem_name)
+
+        subsystem.config['auths.impl.{}.class'.format(plugin_type)] = plugin_class
+        subsystem.save()
+        print("Auth plugin registered.")
+
+
+class AuthPluginDeregisterCLI(pki.cli.CLI):
+    def __init__(self, parent):
+        super(AuthPluginDeregisterCLI, self).__init__(
+            'deregister', 'Deregister auth plugin'
+        )
+
+        self.parent = parent
+
+    def print_help(self):
+        print("Usages: pki-server %s-auth-plugin-deregister <plugin_type>"
+              % self.parent.parent.parent.name)
+        print()
+        print('  -i, --instance <instance ID>       Instance ID (default: pki-tomcat).')
+        print('      --help                         Show help message.')
+        print()
+
+    def execute(self, argv):
+        try:
+            opts, args = getopt.gnu_getopt(argv, 'i:', [
+                'instance=', 'help'])
+
+        except getopt.GetoptError as e:
+            print('ERROR: ' + str(e))
+            self.print_help()
+            sys.exit(1)
+
+        instance_name = 'pki-tomcat'
+        plugin_type = ''
+        for o, a in opts:
+            if o in ('-i', '--instance'):
+                instance_name = a
+
+            elif o == '--help':
+                self.print_help()
+                sys.exit()
+            else:
+                print('ERROR: unknown option ' + o)
+                self.print_help()
+                sys.exit(1)
+        plugin_type = args[0].strip()
+        if not plugin_type:
+            print("ERROR: Plugin type required")
+            sys.exit(1)
+
+        instance = pki.server.PKIInstance(instance_name)
+        if not instance.is_valid():
+            print('ERROR: Invalid instance %s.' % instance_name)
+            sys.exit(1)
+
+        instance.load()
+
+        subsystem_name = self.parent.parent.parent.name
+        subsystem = instance.get_subsystem(subsystem_name)
+        deleted = False
+        for key in list(subsystem.config.keys()):
+            if key.lower().startswith('auths.impl.{}'.format(plugin_type.lower())):
```

Comment:
@edewata I agree with `auths.impl.something.`, i'll fix it.

I used `lower()` to handle user stupidity, I mean, if plugin name is `Plug1012` then sometimes user provided input like `plug1012`. 

What is your thought on this? So I can made the changes accordingly.

Reply:
To my understanding the CA code treats the plugin name as case sensitive. So if we create a CLI that accepts inexact plugin name, it could end up in the configuration file and might break the CA. So I suggest we should just fail the operation if the user doesn't provide the exact plugin name (i.e. don't use `lower()`).